File: problems/896_monotonic_array.rb
=begin
An array is monotonic if it is either monotone increasing or monotone decreasing.

An array A is monotone increasing if for all i <= j, A[i] <= A[j].  An array A is monotone decreasing if for all i <= j, A[i] >= A[j].

Return true if and only if the given array A is monotonic.
=end

# @param {Integer[]} a
# @return {Boolean}
def is_monotonic(a)
    increasing = decreasing = true
    (0...a.length - 1).each do |i|
        if a[i] > a[i+1]
            increasing = false
        elsif a[i] < a[i+1]
            decreasing = false
        end
        if !increasing && !decreasing
            return false
        end
    end
    
    return true
end

File: problems/896_monotonic_array.rs
/*
 * An array is monotonic if it is either monotone increasing or monotone decreasing.

An array nums is monotone increasing if for all i <= j, nums[i] <= nums[j]. An array nums is monotone decreasing if for all i <= j, nums[i] >= nums[j].

Given an integer array nums, return true if the given array is monotonic, or false otherwise.



Example 1:

Input: nums = [1,2,2,3]
Output: true
Example 2:

Input: nums = [6,5,4,4]
Output: true
Example 3:

Input: nums = [1,3,2]
Output: false


Constraints:

1 <= nums.length <= 105
-105 <= nums[i] <= 105
*/

impl Solution {
    pub fn is_monotonic(nums: Vec<i32>) -> bool {
        let mut inc = true;
        let mut dec = true;

        for idx in (1..nums.len()) {
            inc &= nums[idx - 1] <= nums[idx];
            dec &= nums[idx - 1] >= nums[idx];
        }
        return inc || dec;
    }
}

